extensions/mqtt/processors/AbstractMQTTProcessor.h

```diff
@@ -46,150 +44,178 @@ class AbstractMQTTProcessor : public core::Processor {
-    }
-  }
+  /**
+   * Calls disconnect() and releases local MQTT client
+   */
+  void freeResources();
 
-  void onDisconnectionSuccess(MQTTAsync_successData* /*response*/) {
-    logger_->log_info("Successfully disconnected from MQTT broker %s", uri_);
-  }
+  /**
+   * Disconnect from MQTT broker. Synchronously waits until disconnection succeeds or fails.
+   */
+  void disconnect();
+
+  virtual void readProperties(const std::shared_ptr<core::ProcessContext>& context) = 0;
+  virtual void onTriggerImpl(const std::shared_ptr<core::ProcessContext>& context, const std::shared_ptr<core::ProcessSession>& session) = 0;
+  virtual void startupClient() = 0;
+  void setBrokerLimits(MQTTAsync_successData5* response);
+
+  // MQTT static async callbacks, calling their non-static counterparts with context being pointer to "this"
+  static void connectionLost(void *context, char* cause);
+  static void connectionSuccess(void* context, MQTTAsync_successData* response);
+  static void connectionSuccess5(void* context, MQTTAsync_successData5* response);
+  static void connectionFailure(void* context, MQTTAsync_failureData* response);
+  static void connectionFailure5(void* context, MQTTAsync_failureData5* response);
+  static int msgReceived(void *context, char* topic_name, int topic_len, MQTTAsync_message* message);
+
+  // MQTT async callback methods
+  void onConnectionLost(char* cause);
+  void onConnectFinished(MQTTAsync_successData* success_data, MQTTAsync_successData5* success_data_5, MQTTAsync_failureData* failure_data, MQTTAsync_failureData5* failure_data_5);
+  void onDisconnectFinished(MQTTAsync_successData* success_data, MQTTAsync_successData5* success_data_5, MQTTAsync_failureData* failure_data, MQTTAsync_failureData5* failure_data_5);
 
-  void onDisconnectionFailure(MQTTAsync_failureData* response) {
-    logger_->log_error("Disconnection failed from MQTT broker %s (%d)", uri_, response->code);
-    if (response->message != nullptr) {
-      logger_->log_error("Detailed reason for disconnection failure: %s", response->message);
-    }
+  /**
+   * Called if message is received. This is default implementation, to be overridden if subclass wants to use the message.
+   * @param topic topic of message
+   * @param message MQTT message
+   */
+  virtual void onMessageReceived(SmartMessage /*smartmessage*/) {
   }
 
   virtual bool getCleanSession() const = 0;
-  virtual bool startupClient() = 0;
-
-  void freeResources();
-
-  /**
-   * Checks property consistency before connecting to broker
-   */
-  virtual void checkProperties() {
+  virtual bool getCleanStart() const = 0;
+  virtual std::chrono::seconds getSessionExpiryInterval() const = 0;
+  void setMqtt5ConnectOptions(MQTTAsync_connectOptions& conn_opts, MQTTProperties& connect_props, MQTTProperties& will_props) const;
+  virtual void setMqtt5ConnectOptionsImpl(MQTTProperties& /*connect_props*/) const {
```

Comment:
Could the `setMqtt5ConnectOptionsImpl` be removed and the `setMqtt5ConnectOptions` made virtual? In that case we could just override that if needed and call the base class's implementation additionally before adding our own implementation. I could work with `onTrigger` and `onTriggerImpl` as well.

Reply:
Then we must not forget calling base implementation in each derived implementation. If we have them separated we don't have to care about that.

Reply:
True, although maybe it's the naming that doesn't really fit here, maybe it could be something like `setProcessorSpecificMqtt5ConnectOptions`?

Reply:
I have renamed the function to `setProcessorSpecificMqtt5ConnectOptions` in 7e4b4c3ce09969316a03a6a1182e3e540dc645ce